### ISSUE TITLE
Incorrect JOBS table structure on mysql

### ISSUE THREAD

**Issue by sgautrin** [spark-jobserver/spark-jobserver#1187]
**Used Spark Job Server version**

0.8.0 modified for Datastax Enterprise 6.7.2

Jobserver is configured to use mysql backend, and the backend is a mariadb 10.2 galera cluster.

**Actual (wrong) behavior**

The structure of the JOBS table is not as expected, more precisely for the `START_TIME` column: for this column, the `ON UPDATE CURRENT_TIMESTAMP` option is declared on the created table.

**Steps to reproduce**

1. run a docker container with mariadb:10.2, mysql/mysql-server:5.5 or mysql/mysql-server:5.7
2. configure the jobserver to use mysql backend with that server
3. inspect the ceated table structure with `show create table JOBS`

Actual structure of the table ends up being

```sql
CREATE TABLE `JOBS` (
  `JOB_ID` varchar(255) NOT NULL,
  `CONTEXT_NAME` varchar(255) NOT NULL,
  `JAR_ID` int(11) NOT NULL,
  `CLASSPATH` varchar(255) NOT NULL,
  `START_TIME` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  `END_TIME` timestamp NULL DEFAULT NULL,
  `ERROR` text,
  PRIMARY KEY (`JOB_ID`)
);
```

The issue is not reproduced with a mysql/mysql-server:8.0 container.

**Diagnosis**

After noticing the issue, I first digged into the database migrations, and found nothing suspect, the table being created in `spark-jobserver/job-server/src/main/resources/db/mysql/migration/V0_7_0/V0_7_0__init_tables.sql` with

```sql
CREATE TABLE `JOBS` (
  `JOB_ID`          VARCHAR(255)                  NOT NULL PRIMARY KEY,
  `CONTEXT_NAME`    VARCHAR(255)                  NOT NULL,
  `JAR_ID`          INTEGER                       NOT NULL,
  `CLASSPATH`       VARCHAR(255)                  NOT NULL,
  `START_TIME`      TIMESTAMP                     NOT NULL,
  `END_TIME`        TIMESTAMP                     NULL,
  `ERROR`           TEXT                          NULL
);
```

I tested that request on various docker containers (see above) only to get the same incorrect result. Digging around, it seems to be an incorrect behaviour from mysql and mariadb, controlled since mysql 5.6.6 by the[`explicit_defaults_for_timestamp`](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_explicit_defaults_for_timestamp) configuration, which defaults to being disabled up until version 8.0.1 included.

When the configuration is disabled (or prior to mysql 5.6.6)

> The first TIMESTAMP column in a table, if not explicitly declared with the NULL attribute or an explicit DEFAULT or ON UPDATE attribute, is automatically declared with the DEFAULT CURRENT_TIMESTAMP and ON UPDATE CURRENT_TIMESTAMP attributes. 

With these versions, unless the configuration is enabled, declaring a `NOT NULL` timestamp column with result in it getting the `DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP` options.

**Possible solution**

This does not affect as far as I can see the normal behaviour of the jobserver; however, I encountered situations where a job would fail with the jobserver logging an out of heap error, and the jobserver would still list the job as running (essentially what is described in #1161); for those I had imagined setting the ERROR and the END_TIME of the job manually in the table, which resulted in the START_TIME column being updated as well (though I could have “killed” the job which was actually alread failed).

From this structure, the part which is really unwanted is the `ON UPDATE CURRENT_TIMESTAMP` part, as the `DEFAULT CURRENT_TIMESTAMP` makes sense; it would then be possible for mysql to update this column declaration to add the `DEFAULT CURRENT_TIMESTAMP` which would then not automatically add the `ON UPDATE CURRENT_TIMESTAMP`. The alternative would be to make the field nullable which would not really make sense.

**Comment by bsikander** [comment]
For #1161 you already have a PR ready.

For the other issue, we don't need to do anything in SJS right?

**Comment by sgautrin** [comment]
If we did something in SJS, it would be to update the migration scripts for the sql backend.

The thing is, with recent versions of mysql and mariadb, the scripts produce the expected database structure (`START_TIME` is non null and that's it), and the problem affects older versions (but still widely used), where the expected structure is not possible, and for which we would need either

* `NON NULL DEFAULT CURRENT_TIMESTAMP`
* `DEFAULT NULL`

If we did a modification, I guess that would be to change the `START_TIME` column to `START_TIME TIMESTAMP NOT NULL`; there would be most likely other affected columns in other tables added since v0.8.0 like the `CONTEXTS` table which has the same `START_TIME` column (defined identically).

For JOBS table in particular, this is definitely not a big issue, in particular after #1161 is fixed (which was simply a possible reason to want to manually update the table); generally those tables should most likely never be updated “by hand” and the SJS code will always provide the initial `START_TIME` when updating, so the issue is minor and might not warrant using a different table column definitions just for the sake of older mysql/mariadb versions.
